### ISSUE TITLE
Calculate convergence on the Times 1840-1920

### ISSUE THREAD

**Issue by jgonggrijp** [UUDigitalHumanitieslab/ShiCo#7]
In order to determine how many years should be in a model, at a minimum.

**Comment by jgonggrijp** [comment]
Journal for future reference:

 - Cloned the scripts from https://github.com/c-martinez/w2v-convergence/ to Calcatron.
 - Made a Python 3 virtualenv.
 - Made a `settings.py` based on the `settings-template.py`, pointing the `dataDir` to /its/times/TDA_GDA/TDA_GDA_1785-2009/`.
 - Tried installing the requirements, failed because of py3-to-py2 backports `enum34`, `functools32` and `subprocess32`.
 - Removed the backports from the `requirements.txt` and tried again. `pip install` completed successfully.
 - Made an `output` directory inside the source tree.
 - Tried to run `./runConvergenceRange.py --y0=1840 --nYears=20 --outDir=output/`. Didn't work because the module `cPickle` couldn't be found (line 12 of the script).
 - Changed line 12 to `from six.moves import cPickle as pickle` based on https://github.com/Theano/Theano/issues/3739#issuecomment-166651768.
 - Tried to run the script again, failed because the script contained old-fashioned `print` statements. Conclusion: really need Python 2.
 - Threw away the virtualenv and replaced it with a Python 2 virtualenv.
 - Restored the original `requirements.txt` using `git checkout`.
 - Tried to install the requirements again, failed because `Python.h` could not be found.
 - Manually ran `sudo apt install python-dev`.
 - Tried to install the requirements again, success.
 - Tried to run the script again, failed because NLTK couldn't find the `words` corpus.
 - Ran `python -m nltk.downloader words` based on https://stackoverflow.com/a/49048876.
 - Tried to run the script again, failed because the `stopwords` corpus was missing.
 - Downloaded the `stopwords` corpus as well.
 - Tried to run the script again, failed because `randomsentenceloader.py` attempts to `from settings import chunkCache` while the latter isn't defined.

(to be continued)

**Comment by jgonggrijp** [comment]
- Made a `chunks` directory inside the source tree and added this as the `chunkCache` entry to the `settings.py`.
 - Tried to run the script again, success. `output` directory contains a `convergenceRange_1840-1860.pkl` file.
 - Pickle file is empty.

(to be continued)

**Comment by jgonggrijp** [comment]
- Realized I need a preprocessed version of the data rather than the original data. ITS share only contains the original data.
 - Found a preprocessing script in the ShiCo repository, https://github.com/UUDigitalHumanitieslab/ShiCo/blob/master/shico/scripts/times_parsing.py.
 - Cloned the ShiCo repo to Calcatron.
 - Created a `preprocessed` directory inside `/its/times/TDA_GDA`.
 - Created a new Python 2 virtualenv for the preprocessing script.
 - Tried to install the `requirements.txt` of the `shico/scripts` directory, failed because the header files of libxml couldn't be found.
 - Manually ran `sudo apt install libxml2-dev`.
 - Tried installing the requirements again, failed because the compiler still couldn't find `libxml/xmlversion.h`.
 - Manually ran `sudo apt install libxslt1-dev`, based on https://stackoverflow.com/a/15761014.
 - Tried installing the requirements again, failed because `/usr/bin/ld: cannot find -lz`.
 - Manually ran `sudo apt install zlib1g-dev` based on https://stackoverflow.com/a/22883983.

(to be continued)

**Comment by jgonggrijp** [comment]
- Tried installing the requirements again, succes.
 - Generated a `yearFiles.txt` as described on line 85 of the preprocessing script, taking care to remove all years except for 1830-1930.
 - Edited the hardcoded paths in the preprocessing script to match the situation on Calcatron.
 - Ran `python times_parsing.py`, seems to work but takes very long. Working on other things in the meanwhile.

(to be continued)

**Comment by BeritJanssen** [comment]
Thanks for the updates! I also ran into the libxml header issue when I indexed dutchbanking on the cluster, very annoying, and hard to fix. Might document / automate this for future servers? Nothing that has high priority now, of course.
I hope there is enough space on the /its/times virtual disk... Maybe also check with René? Does this mean this preprocessed directory is now going to be available on all servers which use /its/times?

**Comment by jgonggrijp** [comment]
- Preprocessing script crashed on the issue of June 22, 1833 because lxml ran into an element with the name `StartTag` which apparently is invalid. Looking into it.

(to be continued)

**Comment by jgonggrijp** [comment]
@BeritJanssen 

> Thanks for the updates! I also ran into the libxml header issue when I indexed dutchbanking on the cluster, very annoying, and hard to fix. Might document / automate this for future servers? Nothing that has high priority now, of course.

The lxml website does mention this, but of course, you don't read that if it's just one of the dependencies in a `requirements.txt`. I do think we should add this to the ShiCo docs at some point.

> I hope there is enough space on the /its/times virtual disk... Maybe also check with René?

I haven't checked with René, but the Times disk is only half full and the preprocessed data are certainly going to be much smaller than the original data.

> Does this mean this preprocessed directory is now going to be available on all servers which use /its/times?

Yes!

**Comment by jgonggrijp** [comment]
- Inspected the problematic XML file, as well as the one from the preceding day for comparison. The problematic line doesn't actually contain a `<StartTag>` as I thought, but a stray `<` at the end of the line which trips up XML parsers. From the comparison, I could tell this is just a stray character and not a remnant of lost information.
 - Manually removed the stray `<` from the file.
 - Started a `screen` session so I could resume the task, logout from Calcatron, go home and let the script run overnight.
 - Resumed the script, continued successfully past 22. June 1833.
 - Script crashed again on 20. September 1833 because of a mismatch between opening and ending tags `<article>` on line 36190 and `<text.title>` on line 36214.
 - Manually inserted two lines after 36190: `<text>` and then `<text.title>`.
 - Resumed the script again and went home.

(to be continued)

**Comment by jgonggrijp** [comment]
- Script progressed beyond 1833 and then crashed at 14. September 1847 because of an attribute parsing error on line 34503.
 - Line 34503 of the file in question was `<au<L. ALBIANI.</au>`. Manually changed to `<au>L. ALBIANI.</au>`.
 - Resumed the script again and went to do something else.

(to be continued)

**Comment by jgonggrijp** [comment]
- Script crashed at 8. July 1851 because of an opening/ending tag mismatch between `<wd>` on line 75355 and `<p>` on line 75369.
 - Line 75355 of the corrupted file was `<wd pos="472,6099,526,6150">andd>`, manually changed to `<wd pos="472,6099,526,6150">and</wd>`.
 - Resumed the script and went to do something else.

(to be continued)

**Comment by jgonggrijp** [comment]
- Script crashed at 4. November 1854 because of an attribute name parsing error on line 93638.
 - Corrupted line was `<au"JOHN BRIGHT.</au>`, manually changed to `<au>"JOHN BRIGHT.</au>`.
 - Resumed the script and went to have a break.

**Comment by jgonggrijp** [comment]
- Script crashed at 1. July 1862 because of an opening/ending tag mismatch between `wd` on line 39611 and `p` on line 39968.
 - Line 39611 was `<wd pos="1664,5777,2052,5799">Bart,OwnerÂ»ndPetitioner.Take/wd>`, manually changed to `<wd pos="1664,5777,2052,5799">Bart,OwnerÂ»ndPetitioner.Take</wd>`.
 - Resumed the script and went to do something else.

**Comment by jgonggrijp** [comment]
- Script crashed at 6. August 1896 because of a tag mismatch between `ta` on line 127929 and `article` on line 127954.
 - Line 127929 was `<ta>General Post Office, August /ta>`, manually changed to `<ta>General Post Office, August</ta>`.
 - Resumed the script and went home.

**Comment by jgonggrijp** [comment]
- Script crashed at 18. April 1922 because of a tag mismatch between `wd` on line 145505 and `p` on line 145508.
 - Line 145505 was `<wd pos="836,3845,1154,3898">BIRMINGHAM4/wd>`, manually changed to `<wd pos="836,3845,1154,3898">BIRMINGHAM4</wd>`.
 - Resumed the script and went to do something else.

**Comment by jgonggrijp** [comment]
- Preprocessing finished.
 - Closed the `screen` session.
 - Moved the working directory back to the `w2v-convergence` clone.
 - Edited the `settings.py` to make `/its/times/TDA_GDA/preprocessed/` the `dataDir`.
 - Cleared the `output` and `cache` directories.
 - Started a new `screen` session.
 - Activated the virtualenv.
 - Ran `./runConvergenceRange.py --y0=1840 --nYears=20 --outDir=output/`, crashed because NLTK couldn't find the `wordnet` corpus.
 - Ran `python -m nltk.downloader wordnet`.
 - Tried convergence measurement again, seems to take very long. Start 16:55, will check back in 10 minutes.

**Comment by jgonggrijp** [comment]
- Checked back at 17:08, convergence script had progressed from 1840 to 1843. Apparently, measuring the convergence takes at least 4 minutes per year.
 - Went to do something else.

**Comment by jgonggrijp** [comment]
- Checked back at 17:28, 1847 was just finished.
 - Went home.

**Comment by jgonggrijp** [comment]
- Convergence script crashed with the following last lines of output:

```console
Loading cached file: ./cache/1849.pklz
Saving to cache ./cache/1849.pklz...
Building models (1840):
1849 (000001000000),
Traceback (most recent call last):
  File "./runConvergenceRange.py", line 45, in <module>
    measureConvergence(y0, nYears, saveDir)
  File "./runConvergenceRange.py", line 32, in measureConvergence
    yearRange, batchSize, maxSentences)
  File "/home/jgonggrijp/w2v-convergence/convergence.py", line 58, in computeConvergenceOverYearRange
    modelinv = getModelInv(model)
  File "/home/jgonggrijp/w2v-convergence/w2vtransformation.py", line 12, in getModelInv
    return np.linalg.pinv(model.syn0norm)
AttributeError: 'Word2Vec' object has no attribute 'syn0norm'
```

 - Patched the `w2vtransformation.py` based on https://groups.google.com/d/msg/gensim/hlYgjqEVocw/yEq7yhjQEAAJ. Outcommented line 12, uncommented line 11 which already contained the solution.
 - Restarted the convergence measurement script, thankfully it reused the cached files so it could progress immediately to 1850.
 - Convergence script appears to build models in parallel to caching year files.
 - Went to do something else.

**Comment by jgonggrijp** [comment]
- Convergence script crashed with the following last lines of output:

```console
Loading cached file: ./cache/1857.pklz
Saving to cache ./cache/1857.pklz...
1857 (000002000000),
Traceback (most recent call last):
  File "./runConvergenceRange.py", line 45, in <module>
    measureConvergence(y0, nYears, saveDir)
  File "./runConvergenceRange.py", line 32, in measureConvergence
    yearRange, batchSize, maxSentences)
  File "/home/jgonggrijp/w2v-convergence/convergence.py", line 66, in computeConvergenceOverYearRange
    d = testModelConvergence(oldModel, oldModelinv, model, modelinv)
TypeError: testModelConvergence() takes exactly 5 arguments (4 given)
```

 - Invocation of `testModelConvergence` is missing the last argument, the vector size.
 - Searched whether there exists a commit on another branch that fixes this, but this doesn't appear to be the case.
 - Sent an email to Carlos to ask what to do.

**Comment by jgonggrijp** [comment]
- Carlos patched the missing vector size, as well as some outdated gensim client code.
 - Pulled in the changes, discarding my patch of w2vtransformation.py:11-12 because it was superseded by Carlos's patch.
 - Resumed the convergence script, success. `output/convergenceRange_1840-1860.pkl` is 604 bytes.
 - Opened the Python REPL and executed the following lines.

```py
>>> import pickle
>>> convergence = pickle.load(open('output/convergenceRange_1840-1860.pkl'))
>>> convergence
(SortedDict(None, 1000, {2000000: 0.99687576788884602, 2636449: 0.99934428365532879}), SortedDict(None, 1000, {1849: 1000000, 1857: 2000000, 1860: 2636449}), SortedDict(None, 1000, {1000000: 7945, 2000000: 9481, 2636449: 9880}))
```

 - Judging from the documentation and an educated guess, the tuple above is conveying the following information. A model was generated after every millionth sentence and at the end of the year range. The first model was generated for 1840-1849, vector size 7945. Second model 1840-1857, vector size 9481, convergence 0.997. Third model 1840-1860, vector size 9880, convergence 0.999.
 - This tells us 17 years is certainly enough, probably more than enough. Unfortunately, it takes at least two models to determine the convergence, which means we can't determine a lower bound using a batch size of 1 million sentences.
 - Patched `runConvergenceRange.py`, changing the `batchSize` from `1e6` to `1e5`.
 - Ran `./runConvergenceRange.py --y0=1840 --nYears=15 --outDir=output/`, gratefully using the cache files. `output/convergenceRange_1840-1855.pkl` is 2.5KB.
 - Ran similar Python REPL steps as before, producing the following tuple:

```py
(SortedDict(None, 1000, {200000: 0.98466743481059193, 300000: 0.9908352682737821, 400000: 0.99428071346153923, 500000: 0.9970999732338176, 600000: 0.99703888907686145, 700000: 0.99794200052502879, 800000: 0.99779219030514033, 900000: 0.99838594804751035, 1000000: 0.99823335147773717, 1100000: 0.99909925978607628, 1200000: 0.99901016611923121, 1300000: 0.9992330745934489, 1400000: 0.99893929505788714, 1500000: 0.99885838368103275, 1600000: 0.99929347008522029, 1700000: 0.99920699288889392, 1800000: 0.99921171022988253, 1805957: 0.9999999999999919}),
 SortedDict(None, 1000, {1840: 100000, 1842: 300000, 1843: 400000, 1845: 600000, 1846: 700000, 1847: 800000, 1848: 900000, 1849: 1000000, 1850: 1100000, 1851: 1200000, 1852: 1300000, 1853: 1500000, 1854: 1600000, 1855: 1805957}),
 SortedDict(None, 1000, {100000: 1840, 200000: 2211, 300000: 2517, 400000: 2756, 500000: 2897, 600000: 3056, 700000: 3178, 800000: 3321, 900000: 3435, 1000000: 3570, 1100000: 3642, 1200000: 3725, 1300000: 3792, 1400000: 3889, 1500000: 4000, 1600000: 4072, 1700000: 4155, 1800000: 4242, 1805957: 4242}))
```

 - At first sight, the numbers are too good to be true; the models appear to reach convergence already after 2e5 sentences (less than three years). However, this made me realize that numbers derived from different batch sizes are incomparable. Especially telling is that the vector sizes are much smaller. I cannot explain why the vector size for 1.8e6 sentences (1855) is only 4242, when it was 7945 for 1e6 sentences (1849) before.
 - Patched `runConvergenceRange.py` one more time, changing the `batchSize` to `5e5`, hoping that this would at least make the vector sizes more comparable.
 - Ran `./runConvergenceRange.py --y0=1840 --nYears=17 --outDir=output/`. `output/convergenceRange_1840-1857.pkl` is 827 bytes.
 - Opened the pickle file in the Python REPL, producing the following tuple.

```py
(SortedDict(None, 1000, {1000000: 0.99495217206551012, 1500000: 0.99765601904936019, 2000000: 0.99823252577862187, 2114912: 0.99995114618184711}),
 SortedDict(None, 1000, {1845: 500000, 1849: 1000000, 1853: 1500000, 1857: 2114912}),
 SortedDict(None, 1000, {500000: 5259, 1000000: 6347, 1500000: 7022, 2000000: 7632, 2114912: 7650}))
```

 - Conclusion: 10 years *might* be enough, but this is impossible to confirm with certainty given the low number of sentences per year. 17 years is certainly enough.

**Comment by jgonggrijp** [comment]
- Reopened, because I realized it would be informative to have convergence measurements of a few more year ranges.
 - Un-patched the `runConvergenceRange.py`, resetting the `batchSize` to `1e6`.
 - Started `./runConvergenceRange.py --y0=1870 --nYears=20 --outDir=output/`.
 - No caches available, so went to do something else.

**Comment by jgonggrijp** [comment]
- Convergence on 1870-1890 finished, pickle file is 888 bytes, content below.

```py
(SortedDict(None, 1000, {2000000: 0.99739326176089715, 3000000: 0.99848624118043883, 4000000: 0.99903969464228759, 4935706: 0.99938535525288918}),
 SortedDict(None, 1000, {1875: 1000000, 1880: 2000000, 1883: 3000000, 1887: 4000000, 1890: 4935706}),
 SortedDict(None, 1000, {1000000: 8032, 2000000: 9299, 3000000: 10226, 4000000: 10931, 4935706: 11439}))
```

 - Conclusion: 10 years is clearly enough in this range. Vectors are large.
 - Started `./runConvergenceRange.py --y0=1905 --nYears=15 --outDir=output/`.
 - Went to do something else.

**Comment by jgonggrijp** [comment]
- Convergence on 1905-1920 finished, pickle file is 1.2 KB, content below.

```py
(SortedDict(None, 1000, {2000000: 0.99599745238260651, 3000000: 0.9982040655648996, 4000000: 0.99820539852190338, 5000000: 0.99884741921501607, 6000000: 0.99928363485605476, 6769275: 0.99966717468698563}),
 SortedDict(None, 1000, {1908: 1000000, 1911: 2000000, 1913: 3000000, 1916: 4000000, 1918: 5000000, 1919: 6000000, 1920: 6769275}),
 SortedDict(None, 1000, {1000000: 7194, 2000000: 8851, 3000000: 9875, 4000000: 11178, 5000000: 12202, 6000000: 12945, 6769275: 13319}))
```

 - Conclusion: 6 years is already enough.
 - Started `./runConvergenceRange.py --y0=1850 --nYears=20 --outDir=output/` for good measure.
 - Went to do something else.

**Comment by jgonggrijp** [comment]
- Convergence on 1850-1870 finished, pickle file is 759 bytes, content below.

```py
(SortedDict(None, 1000, {2000000: 0.99714172988061245, 3000000: 0.99837624471514874, 3547842: 0.99962511929281617}),
 SortedDict(None, 1000, {1857: 1000000, 1863: 2000000, 1868: 3000000, 1870: 3547842}),
 SortedDict(None, 1000, {1000000: 7805, 2000000: 9132, 3000000: 10110, 3547842: 10365}))
```

 - Conclusion: 13 years is enough in this range.

Summary:

Starting year | Upper bound on convergence
---|---
1840 | 17
1850 | 13
1870 | 10
1905 | 6
